### ISSUE TITLE
Failed tests on Debian & Ubuntu with 6.9.6.6 and 3.4.3~rc1

### ISSUE THREAD

**Issue by nacc** [mkoppanen/imagick#187]
Hello,

You were very helpful to me in the past, and I was wondering if you would be able to help run a travis test run with the $subject versions of imagemagick and php-imagick respectively? We are seeing a few failures not seen by the current CI and I'm wondering if they are upstream fixes that we need to backport for imagemagick. Note that two failures are confusingly about a method not being defined when it clearly is in the source. Log at : https://objectstorage.prodstack4-5.canonical.com/v1/AUTH_77e2ada1e7a84929a74ba3b87153c0ac/autopkgtest-zesty/zesty/armhf/p/php-imagick/20161129_202720_72eed@/log.gz [while armhf, i see the same on amd64 but our builders are backed up ].

**Comment by Danack** [comment]
I've go on my schedule to do some work for Imagick on Thursday - I can have a look then.


In the meantime - can you do some debugging. I suspect that something is weird with your CI machine. The autoOrient function should be there as you're compiling against >= 0x692 which is the only thing it checks for......but apprently it's not actually there.


I think doing something like:

#error MagickLibVersion is ##MagickLibVersion

in one of the C files, should get the version.

**Comment by nacc** [comment]
You were very helpful again! So the issue was the Ubuntu package hadn't been rebuilt against the updated ImageMagick package. However, I'm still seeing the following two issues. The first seems like an off-by-one somewhere? Would that be in ImageMagick or php-imagick?

```
There was 1 failure:

1) /tmp/autopkgtest.BDn7KT/build.2bw/php-imagick-3.4.3~rc1/imagick-3.4.3RC1/tests/256_Imagick_exportImagePixels_basic.phpt
Failed asserting that format description matches text.
--- Expected
+++ Actual
@@ @@
+pixel type 2 has incorrect values. They should be 0/255, 1/255, 2/255....or the scaled equivalent
+array(10) {
+  [0]=>
+  float(0)
+  [1]=>
+  float(0.0039063096055543)
+  [2]=>
+  float(0.0078126192111086)
+  [3]=>
+  float(0.011718928816663)
+  [4]=>
+  float(0.015625238422217)
+  [5]=>
+  float(0.019531548027771)
+  [6]=>
+  float(0.023437857633326)
+  [7]=>
+  float(0.02734416723888)
+  [8]=>
+  float(0.031250476844434)
+  [9]=>
+  float(0.035156786449989)
+}
+pixel type 3 has incorrect values. They should be 0/255, 1/255, 2/255....or the scaled equivalent
+array(10) {
+  [0]=>
+  float(0)
+  [1]=>
+  float(0.0039063096046448)
+  [2]=>
+  float(0.0078126192092896)
+  [3]=>
+  float(0.011718928813934)
+  [4]=>
+  float(0.015625238418579)
+  [5]=>
+  float(0.019531548023224)
+  [6]=>
+  float(0.023437857627869)
+  [7]=>
+  float(0.027344167232513)
+  [8]=>
+  float(0.031250476837158)
+  [9]=>
+  float(0.035156786441803)
+}
+pixel type 5 has incorrect values. They should be 0/255, 1/255, 2/255....or the scaled equivalent
+array(10) {
+  [0]=>
+  int(0)
+  [1]=>
+  int(16777472)
+  [2]=>
+  int(33554944)
+  [3]=>
+  int(50332416)
+  [4]=>
+  int(67109888)
+  [5]=>
+  int(83887360)
+  [6]=>
+  int(100664832)
+  [7]=>
+  int(117442304)
+  [8]=>
+  int(134219776)
+  [9]=>
+  int(150997248)
+}
+pixel type 6 has incorrect values. They should be 0/255, 1/255, 2/255....or the scaled equivalent
+array(10) {
+  [0]=>
+  int(0)
+  [1]=>
+  int(256)
+  [2]=>
+  int(512)
+  [3]=>
+  int(768)
+  [4]=>
+  int(1024)
+  [5]=>
+  int(1280)
+  [6]=>
+  int(1536)
+  [7]=>
+  int(1792)
+  [8]=>
+  int(2048)
+  [9]=>
+  int(2304)
+}
+pixel type 7 has incorrect values. They should be 0/255, 1/255, 2/255....or the scaled equivalent
+array(10) {
+  [0]=>
+  int(0)
+  [1]=>
+  int(256)
+  [2]=>
+  int(512)
+  [3]=>
+  int(768)
+  [4]=>
+  int(1024)
+  [5]=>
+  int(1280)
+  [6]=>
+  int(1536)
+  [7]=>
+  int(1792)
+  [8]=>
+  int(2048)
+  [9]=>
+  int(2304)
+}
+pixel type 4 has incorrect values. They should be 0/255, 1/255, 2/255....or the scaled equivalent
+array(10) {
+  [0]=>
+  int(0)
+  [1]=>
+  int(16777472)
+  [2]=>
+  int(33554944)
+  [3]=>
+  int(50332416)
+  [4]=>
+  int(67109888)
+  [5]=>
+  int(83887360)
+  [6]=>
+  int(100664832)
+  [7]=>
+  int(117442304)
+  [8]=>
+  int(134219776)
+  [9]=>
+  int(150997248)
+}
 Ok
```

**Comment by Danack** [comment]
> Would that be in ImageMagick or php-imagick?

It will be in ImageMagick. php-imagick doesn't really do that much.

From your tests, you have a value of 0.0039063096055543  - this should be 1/255, which is 0.003921568627451

_Highly coincidentally_ the value of 0.0039063096055543 can be obtained by doing (1 / 256) * (65536 / 65535). I'm guessing you're running ImageMagick compiled with 16bit values per pixel as 2**16 = 65536.

Which might imply a double bug; both the normalised values going 0->256 and 0->65536 would be incorrect. The real range is meant to be 0->255, and 0->65535.

There is a chance the tests in PHP Imagick are incorrect....but at the least the behaviour has changed from what it was before.

Someone will probably need to write a C example to demonstrate this, as the ImageMagick guys like to have simple examples to repro issues.

**Comment by nacc** [comment]
>> Would that be in ImageMagick or php-imagick?
> It will be in ImageMagick. php-imagick doesn't really do that much.

Ok, that is what I suspected. Thank you for confirming.

> Highly coincidentally the value of 0.0039063096055543 can be obtained by doing (1 / 256) * (65536 > / 65535). I'm guessing you're running ImageMagick compiled with 16bit values per pixel as 2**16 =
> 65536

Yes, that's correct, I believe (q16, as on Debian). Is there a trivial way for me to confirm this?

> Which might imply a double bug; both the normalised values going 0->256 and 0->65536 would be
> incorrect. The real range is meant to be 0->255, and 0->65535.

Right, that's what it appears to be (off by one in both ranges).

Also, I just tested changing

`$imagick->newPseudoImage(256, 256, "gradient:black-white");`
to
`$imagick->newPseudoImage(255, 255, "gradient:black-white");`

In the testcase, and it passes. That seems confusing to me, as those values only define the dimensions of the image, right?

> Someone will probably need to write a C example to demonstrate this to be able to report it at
> https://github.com/ImageMagick/ImageMagick, as the ImageMagick guys like to have simple
> examples to repro issues.

Sadly, I'm not actually a user of ImageMagick, just trying to get an update rolled out to Ubuntu 17.04 :/ I will do my best to get something written up, though.

-Nish

**Comment by nacc** [comment]
@Danack can you confirm the change to the testcase is just a red herring?

**Comment by nacc** [comment]
@Danack, also, I'm not sure if your CI environment allows for this, but I think you were able to do this in the past, using upstream imagemagick and php-imagick, could you test against 6.9.9-6 ? I see a run for 6.9.2-0 (https://travis-ci.org/mkoppanen/imagick/jobs/179743966) -- if you are able to reproduce it, I could then try to bisect, at least.

**Comment by Danack** [comment]
> can you confirm the change to the testcase is just a red herring?

No, it's a clue, and should be a useful thing to test against.


I will do, but you can also do that in a fork at https://github.com/mkoppanen/imagick/blob/master/.travis.yml#L17-L28

You can also get it to compile against commit shas, which should also be useful for finding the exact commit.

**Comment by nacc** [comment]
@Danack Great, thanks! It does seem like the Launchpad mirror that the GitHub Travis is using doesn't have the most recent releases. I'm talking to the LP folks here to figure out why.

**Comment by nacc** [comment]
Reproduced the test failure in the meanwhile, using the git commit: https://travis-ci.org/nacc/imagick/builds/180778172

**Comment by nacc** [comment]
Bisected it down, thanks for all your help!

https://travis-ci.org/nacc/imagick/builds/180791451 indicates the regression in php-imagick happens with https://github.com/ImageMagick/ImageMagick/commit/d75ed24792891c10837da941adea162073093dda.

I see that it's a bugfix (based upon the upstream discussion link) in the gradient code, but I wonder if it shifted an assumption in the testcase?

**Comment by Danack** [comment]
You seem to think this is an urgent thing that needs to be fixed as soon as possible. What's the source of your urgency?

I think the current tests are correct. Even if they aren't I'm not going to commit a change to make them pass without talking to the ImageMagick guys to find out what has changed.

To explain the test it's making an image 256 pixel wide, aka columns 0-255. The image should be a gradient from black (0,0,0) to white(255, 255, 255). Each column should be incrementing by 1.

Apparently it's not quite....but it's really quite close. 

Expected: 0.003921568627451
Actual: 0.0039063096055543

Only people looking at unit tests will notice the difference, so I'm not going to hack around this at short notice.

**Comment by nacc** [comment]
@Danack, sorry for giving a false impression! I tried to make this clear in my PR: "so is not suitable for merging (yet)."

Reasons for providing a PR that I don't expect to be pulled:

1) I want to put code out there related to this issue, I could just have referred to the issue in the commit message (as I did) and I didn't realize GH would note that here (I thought it needed the PR).
2) I am travelling for a week and may not be as responsive as before.
3) Both Ubuntu and Debian's autopkgtests for php-imagick have been failing. For Debian, this doesn't block anything, but for Ubuntu it means we can't update imagemagick in 17.04 until the testing issue has been fixed.

In any case, I want to express clearly that I appreciate your help, I'm not trying to pressure you to merge anything you are not comfortable and that absolutely the right thing to do is to talk to ImageMagick upstream to determine what, if anything, has changed, and why. I've tried to provide as much information as possible in order to help that discussion. But, at the same time, I'm a disinterested party in that discussion itself, as I don't use ImageMagick or php-imagick :)
